### PR TITLE
feat: add filter in elasticsearch backend

### PR DESCRIPTION
**Pull request by jemmyshin**
closes: https://github.com/jina-ai/docarray/issues/374

### CODE REVIEW

File: docarray/array/storage/elastic/find.py [hunk 1]
```diff
@@ -31,21 +29,51 @@
 
 
 class FindMixin(BaseFindMixin):
-    def _find_similar_vectors(self, query: 'ElasticArrayType', limit=10):
+
+    def _find_similar_vectors(self, query: 'ElasticArrayType', limit=10, **kwargs,
```
Comment (numb3r3) [inline docarray/array/storage/elastic/find.py:33]: add the parameter `filter_filed`, `filter_values` in the function signature, and add docstring.

File: docarray/array/storage/elastic/find.py [hunk 2]
```diff
@@ -31,21 +29,51 @@
...
-                'k': limit,
-                'num_candidates': 10000,
-            },
-        )
+        filter_field = kwargs["filter_field"]
+        filter_values = kwargs["filter_values"]
+        if filter_field and filter_values:
+            if filter_field != 'id':
+                raise ValueError(f'vector search in elastic search only support filter '
+                                 f'on id,got: {filter_field}')
```
Comment (numb3r3) [inline docarray/array/storage/elastic/find.py:45]: add space after ","

File: docarray/array/storage/elastic/find.py [hunk 3]
```diff
@@ -31,21 +29,51 @@
...
+            q = {
+                "script_score": {
+                    "query": {
+                        "bool": {
+                            "filter": {filter_field: {"values": filter_values}}
+                        },
+                    },
+                    "script": {
+                        "source": "cosineSimilarity(params.query, 'embedding') + 1.0",
```
Comment (numb3r3) [inline docarray/array/storage/elastic/find.py:54]: should support other metrics? euclidean?

Comment (numb3r3) [reply docarray/array/storage/elastic/find.py:54]: for me, it looks like:
```
sub_docs = docs[sub_fileds]

query_docs.matches(sub_docs)
```

Comment (numb3r3) [reply docarray/array/storage/elastic/find.py:54]: The matched `score` values assignment is missing? (refer to line 82)

File: docarray/array/storage/elastic/find.py [hunk 4]
```diff
@@ -58,68 +86,96 @@ def _find_similar_vectors(self, query: 'ElasticArrayType', limit=10):
...
 
+        :param filter_field: inside **kwargs, which field to filter when searching.
+        :param filter_values: inside **kwargs, candidate values for `filter_field`, the values of `filter_field` for results
+            must be in `filter_values`.
+
         :return: DocumentArray containing the closest documents to the query if it is a single query, otherwise a list of DocumentArrays containing
            the closest Document objects for each of the queries in `query`.
         """
 
+        query = {
```
Comment (numb3r3) [inline docarray/array/storage/elastic/find.py:105]: This mapping should be optional?

**Review by numb3r3 — CHANGES_REQUESTED** [review]

File: docarray/array/storage/elastic/find.py (hunk 1 above)
Comment (numb3r3) [reply docarray/array/storage/elastic/find.py:33]: reuse `filter` parameter in `find()` method

**Comment by numb3r3** [comment]
and add unittest

**Comment by numb3r3** [comment]
@jemmyshin don't use `merge`, please use the following command to rebase your PR:

```
git pull --rebase origin main
```

File: docarray/array/mixins/find.py [hunk 5]
```diff
@@ -132,9 +132,9 @@ def find(
 
         from ... import Document, DocumentArray
 
-        if isinstance(query, dict):
+        if isinstance(query, dict) or isinstance(query[0], dict):
             if filter is None:
-                return self._filter(query)
+                return self._filter(query, **kwargs)
```
Comment (numb3r3) [inline docarray/array/mixins/find.py:137]:
```
self._filter(query, limit=limit)
```

File: docarray/array/mixins/find.py [hunk 6]
```diff
@@ -132,9 +132,9 @@ def find(
 
         from ... import Document, DocumentArray
 
-        if isinstance(query, dict):
+        if isinstance(query, dict) or isinstance(query[0], dict):
```
Comment (numb3r3) [inline docarray/array/mixins/find.py:135]: rever this commit, it does not make sense to accept batch query.

File: docarray/array/mixins/find.py [hunk 7]
```diff
@@ -255,9 +255,8 @@ def _find(
         raise NotImplementedError
 
     def _filter(
-        self,
-        query: Dict,
-    ) -> 'DocumentArray':
+        self, query: Union[Dict, List[Dict]], **kwargs
+    ) -> Union['DocumentArray', List['DocumentArray']]:
```
Comment (numb3r3) [inline docarray/array/mixins/find.py:259]: revert this commit

File: docarray/array/storage/elastic/find.py [hunk 8]
```diff
@@ -123,3 +131,29 @@ def _find(
             query = query.reshape((num_rows, -1))
 
         return [self._find_similar_vectors(q, limit=limit) for q in query]
+
+    def _find_with_filter(self, query: Dict, limit: int = 10):
```
Comment (numb3r3) [inline docarray/array/storage/elastic/find.py:135]: The function signature: query: ArrayLike, filter: Dict

**Review by numb3r3 — CHANGES_REQUESTED** [review]

File: docarray/array/storage/elastic/find.py [hunk 9]
```diff
@@ -123,3 +131,24 @@ def _find(
...
```
Comment (numb3r3) [inline docarray/array/storage/elastic/find.py:135]: limit should optional

File: docarray/array/storage/elastic/find.py [hunk 10]
```diff
@@ -123,3 +131,24 @@ def _find(
...
+
+        da = DocumentArray()
+        for result in list_of_hits[:limit]:
+            doc = Document.from_base64(result['_source']['blob'])
+            doc.scores['score'] = NamedScore(value=result['_score'])
+            da.append(doc)
+
+        return da
+
+    def _filter(self, query: Dict, limit: int = 10) -> 'DocumentArray':
```
Comment (numb3r3) [inline docarray/array/storage/elastic/find.py:152]: limit is optional

File: docarray/array/storage/elastic/find.py (hunk 9 above)
Comment (numb3r3) [reply docarray/array/storage/elastic/find.py:135]: add docstr

**Review by numb3r3 — CHANGES_REQUESTED** [review]

File: docarray/array/storage/elastic/find.py [hunk 11]
```diff
@@ -105,11 +116,8 @@ def _find(
         **kwargs,
     ) -> List['DocumentArray']:
         """Returns approximate nearest neighbors given a batch of input queries.
-
         :param query: input supported to be stored in Elastic. This includes any from the list '[np.ndarray, tensorflow.Tensor, torch.Tensor, Sequence[float]]'
         :param limit: number of retrieved items
-        :param filter: filter query used for pre-filtering
```
Comment (alaeddine-13) [inline docarray/array/storage/elastic/find.py:111]: why is this removed ?

File: docarray/array/storage/elastic/find.py [hunk 12]
```diff
@@ -123,3 +131,24 @@ def _find(
...
+        for result in list_of_hits[:limit]:
+            doc = Document.from_base64(result['_source']['blob'])
+            doc.scores['score'] = NamedScore(value=result['_score'])
+            da.append(doc)
+
+        return da
+
+    def _filter(self, query: Dict, limit: int = 10) -> 'DocumentArray':
+
+        return self._find_with_filter(query, limit=limit)
```
Comment (alaeddine-13) [inline docarray/array/storage/elastic/find.py:154]: can't we just put the implementation of _find_with_filter directly inside _filter ?

File: docarray/array/storage/elastic/find.py (hunk 11 above)
Comment (jemmyshin) [reply docarray/array/storage/elastic/find.py:111]: added

File: docarray/array/storage/elastic/find.py (hunk 12 above)
Comment (jemmyshin) [reply docarray/array/storage/elastic/find.py:154]: this is just aligned with the interface since we have this `_filter` function in other backends and also in `mixin/find.py`.

File: tests/unit/array/mixins/test_find.py [hunk 13]
```diff
@@ -344,3 +344,28 @@ def test_unsupported_pre_filtering(storage, start_storage):
...
+    'storage, config',
+    [
+        ('elasticsearch', {'n_dim': 32, 'index_text': False}),
+    ],
+)
+def test_elastic_id_filter(storage, config):
+    da = DocumentArray(storage=storage, config=config)
+    da.extend(
+        [
+            Document(id=f'r{i}', embedding=np.random.rand(1, 32)) for i in range(50)
```
Comment (numb3r3) [inline tests/unit/array/mixins/test_find.py:359]: the embedding shape should be `(32, )`, rather than `(1, 32)`

File: tests/unit/array/mixins/test_find.py [hunk 14]
```diff
@@ -344,3 +344,28 @@ def test_unsupported_pre_filtering(storage, start_storage):
...
+    )
+    id_list = [np.random.choice(50, 10, replace=False) for _ in range(10)]
+
+    for id in id_list:
+        query = {
+                "bool": {
+                    "filter": {"ids": {"values": id}}
+                }
+        }
+        result = da.find(query=query)
```
Comment (numb3r3) [inline tests/unit/array/mixins/test_find.py:370]: also test whether `limit` can works as well

**Review by numb3r3 — CHANGES_REQUESTED** [review]

**Comment by numb3r3** [comment]
@jemmyshin please fix the issue related with code formating

**Review by numb3r3 — APPROVED** [review]
LGTM

File: tests/unit/array/mixins/test_find.py [hunk 15]
```diff
@@ -344,3 +344,25 @@ def test_unsupported_pre_filtering(storage, start_storage):
...
+    id_list = [np.random.choice(50, 10, replace=False) for _ in range(3)]
+
+    for id in id_list:
+        id = list(map(lambda x: str(x), id))
+        query = {
+            "bool": {"filter": {"ids": {"values": id}}},
+        }
+        result = da.find(query=query, limit=limit)
+        assert all([r.id in id for r in result])
+        assert len(result) == limit
```
Comment (alaeddine-13) [inline tests/unit/array/mixins/test_find.py:368]: should be in test_filter instead

**Review by JoanFM — APPROVED** [review]

**Review by davidbp — APPROVED** [review]
LGTM
